### PR TITLE
optimize setLanguage, setScript, shapeWithTrace

### PR DESCRIPTION
**Pull request by chearon**
I'm seeing `createCString` as a dominator in a project I'm working on. Nodejs text encoding/decoding is slow: nodejs/node#39879. Browsers don't show much difference with this change.

Here's how I profiled it in harfbuzzjs:

```diff
diff --cc examples/hbjs.example.js
index b7718e9,b7718e9..75b4c3a
--- a/examples/hbjs.example.js
+++ b/examples/hbjs.example.js
@@@ -7,8 -7,8 +7,15 @@@ function example(hb, fontBlob, text) 
    font.setScale(1000, 1000); // Optional, if not given will be in font upem
  
    var buffer = hb.createBuffer();
++  console.time('c strings');
++  for (let i = 0; i < 1000; i++) {
++    buffer.setLanguage('en');
++    buffer.setScript('Latn');
++  }
++  console.timeEnd('c strings');
    buffer.addText(text || 'abc');
--  buffer.guessSegmentProperties();
++  buffer.setDirection('ltr');
++  buffer.setClusterLevel(1);
    // buffer.setDirection('ltr'); // optional as can be set by guessSegmentProperties also
    hb.shape(font, buffer); // features are not supported yet
    var result = buffer.json(font);
```

### CODE REVIEW

File: hbjs.js [hunk 1]
```diff
@@ -238,14 +238,15 @@ function hbjs(instance) {
...
 
-  var utf8Encoder = new TextEncoder("utf8");
-  function createCString(text) {
-    var bytes = utf8Encoder.encode(text);
-    var ptr = exports.malloc(bytes.byteLength);
-    heapu8.set(bytes, ptr);
+  function createAsciiString(text) {
+    var ptr = exports.malloc(text.length + 1);
+    for (let i = 0; i < text.length; ++i) {
+      heapu8[ptr + i] = text.charCodeAt(i);
```
Comment (khaledhosny) [inline hbjs.js:244]: This this takes user input, I’d fo some error checking here that the code point is indeed in the ASCII range.

**Comment by khaledhosny** [comment]
Thanks, the use of TextEncoder here always felt like an overkill.

File: hbjs.js [hunk 2]
```diff
@@ -329,7 +330,7 @@ function hbjs(instance) {
       * @param {string} script: The buffer script
       */
       setScript: function (script) {
-        var str = createCString(script);
+        var str = createAsciiString(script);
```
Comment (khaledhosny) [inline hbjs.js:333]: It probably does not matter in practice, but you can turn the string into a tag yourself and use hb_script_from_iso15924_tag()

Comment (chearon) [reply hbjs.js:339]: I may try that and open a followup. I also wonder if there would be a big difference if you can pass an `hb_tag` directly to `setScript` and `setLanguage` (so it accepts `string | number`).

**Pull request merged**